stl/src/winapisupp.cpp

```diff
@@ -548,6 +564,7 @@ extern "C" PVOID __KERNEL32Functions[eMaxKernel32Function] = {0};
 
 static int __cdecl initialize_pointers() {
     HINSTANCE hKernel32 = GetModuleHandleW(L"kernel32.dll");
+    HINSTANCE hSynch    = GetModuleHandleW(L"api-ms-win-core-synch-l1-2-0.dll");

```

Comment:
(We can `GetModuleHandle` on `kernel32` because we always know it is loaded first; we don't know that for `api-ms-win-core-synch-l1-2-0`)

Reply:
Honestly I'm concerned about this too.

`LoadLibrary[Ex]` call can be used, if `GetModuleHandle` is not guaranteed. The main issue to me here is the module name itself. 

Despite the documentation, is is not in `kernel32.dll`. It is available in `KernelBase,dll`, but `Synchronization.lib` seem to point to `api-ms-win-core-synch-l1-2-0`...

Reply:
`api-ms-win-core-synch-l1-2-0` is the right place -- we can unconditionally use the ApiSet since this is a Windows 8.1+ API, and Windows 8.1 is always ApiSet aware. But we can't necessarily assume that module is already loaded like we can with kernel32, so this would need a LoadLibrary and FreeLibrary pair -- unless there's something about ApiSets that lets the GetModuleHandle call work. @stwish-msft is the relative expert on the behavior of ApiSets to me so I hope he can comment.

Reply:
My preference would be then `GetModuleHandleEx` with `GET_MODULE_HANDLE_EX_FLAG_PIN` on that ApiSet.  Alternatives I see are either referencing the DLL on its API calls, or catching access violations if some late `atexit` callback decides to notify atomics. Albeit AVs are for hypothetical case when `FreeLibrary` really unloads the DLL. I'm sure `KernelBase.dll` is already pinned anyway.

Reply:
>GetModuleHandleEx with GET_MODULE_HANDLE_EX_FLAG_PIN 

That doesn't fix the problem that the module might not be loaded in the first place. We also can't `_PIN` as that prevents the standard library from being cleanly unloaded.

Like I said I would refrain from touching this until we have someone who actually understands apisets here to comment.

Reply:
Ok, will wait for ApiSet expert.
In mean time, I'd also propose /DELAYLOAD as another possible approach. Especially that it will go to another DLL anyway, so current DLLs will not be altered. Any reason existing APIs are `GetProcAddress`'ed, instead of delay load?

Reply:
My understanding of delayload is that this is a performance optimization thing not a correctness thing (that is, it explodes if you try to touch one of those functions and they don't exist) but it's a feature I've never used myself so I won't say for sure there.

Reply:
/DELAYLOAD can be used to avoid LoadLibrary / GetProcAdrress:
> This relieves you of the need to use the Windows SDK functions LoadLibrary and GetProcAddress to implement DLL delayed loading.

(from [Linker Support for Delay-Loaded DLLs](https://docs.microsoft.com/en-us/cpp/build/reference/linker-support-for-delay-loaded-dlls?view=vs-2019))

> it explodes if you try to touch one of those functions and they don't exist

This is really a default behavior, but it can be customized by providing a fallback. Also "exploded" means "throws an exception", but you can catch it (and remember that this attempt throws). See [Error Handling and Notification](https://docs.microsoft.com/en-us/cpp/build/reference/error-handling-and-notification?view=vs-2019). The behavior can be even deeper customized if [own helper is implemented instead of taking delayimp.lib](https://docs.microsoft.com/en-us/cpp/build/reference/delayload-delay-load-import?view=vs-2019)

Finally, you can avoid touching the bomb by calling `VerifyVersionInfo` to check for Windows 8. The difference from static import is that executable would run (or DLL would load) at all without having some functions available.

> a performance optimization thing

Maybe, if comparing against `GetProcAddress`. Import table resolution should be faster than `GetProcAddress` of multiple functions, since it would take advantage of both export and import names sorted, will not search each name from the beginning. Also it may take loader lock just once. (Delay imports is a directory in executable file format, similar to imports; it is not just call of `LoadLibrary` / `GetProcAddress` by a runtime helper)

Reply:
Ok, /DELAYLOAD is apparently not universal solution, therefore not a solution.

Whereas it works to specify /DELAYLOAD:"API-MS-WIN-CORE-SYNCH-L1-2-0.DLL" for a DLL, it does not work for a static library. In case of static library, switch is specified to the target binary instead.

Thus, it will not be compatible with /MT or /MTd, as it is not legal to ask user to add switches except /MT or /MTd.

Still waiting for response on ApiSet anyway.

Reply:
Has this ever been fully discussed? I see it was merged as is but am still unsure whether this is correct behaviour.

Reply:
Yes. Since we are importing `SRWLOCK` (currently) or `CTRITICAL_SECTION` (previously), synchronization library is loaded, so `GetModuleHandle` is safe. `API-MS-WIN-CORE-SYNCH-L1-2-0.DLL` is a forwarder to synchronization library, either `Kernel32.dll` or `KernelBase.dll`.